### PR TITLE
feat: `record_batch!` macro

### PR DESCRIPTION
**Pull request by ByteBaker**
closes: #6553

### CODE REVIEW

**Comment by ByteBaker** [comment]
For your review @alamb

File: arrow-array/src/record_batch.rs [hunk 1]
```diff
@@ -79,6 +79,76 @@ pub struct RecordBatch {
...
+/// Example:
+///
+/// ```rust
+///
+/// use arrow_array::create_array;
+///
+/// let array = create_array!(Int32, [1, 2, 3, 4, 5]);
+/// ```
+#[macro_export]
+macro_rules! create_array {
```
Comment (alamb) [inline arrow-array/src/record_batch.rs:94]: @ByteBaker  would you be willing to extend this macro to cover the other variants of 

https://docs.rs/arrow/latest/arrow/datatypes/enum.DataType.html


Maybe we could simply start with "non nested" types (e.g. no list, dictionary, REE, etc) ?

**Review by alamb — APPROVED** [review]
Thank you @ByteBaker  -- this looks great to me.

I think it might also be nice to put a link to this macro on other docs

https://docs.rs/arrow/latest/arrow/record_batch/struct.RecordBatch.html#method.try_from_iter

And also to add an example to the the RecordBatch docs themselves https://docs.rs/arrow/latest/arrow/record_batch/struct.RecordBatch.html

cc @timsaucer

**Comment by timsaucer** [comment]
Thank you for doing this!

File: arrow-array/src/record_batch.rs (hunk 1 above)
Comment (ByteBaker) [reply arrow-array/src/record_batch.rs:94]: Certainly! Lemme get back with the updates soon.

**Comment by ByteBaker** [comment]
@alamb you're right. I was just lazy and did the bare-minimum. Lemme see how I can further extend this to cover more types and add necessary docs wherever applicable. I'll be back soon.

**Comment by alamb** [comment]
Marking as draft as the CI is not passing

**Comment by ByteBaker** [comment]
@alamb sorry but I've been somewhat occupied the past several weeks. There's some work I've done but I don't think it's quite ready yet.

File: arrow-array/src/record_batch.rs (hunk 1 above)
Comment (ByteBaker) [reply arrow-array/src/record_batch.rs:94]: @alamb Yes, I had started working on this, but I'm not sure if it's anywhere close to complete. Following data types are presently supported.

`Null`,
`Boolean`,
`Int8`,
`Int16`,
`Int32`,
`Int64`,
`UInt8`,
`UInt16`,
`UInt32`,
`UInt64`,
`Float16`,
`Float32`,
`Float64`,
`Utf8`,
`Utf8View`,
`LargeUtf8`,
`Binary`,
`LargeBinary`,
`IntervalDayTime`,
`IntervalYearMonth`,
`Second`,
`Millisecond`,
`Microsecond`,
`Nanosecond`,
`Second32`,
`Millisecond32`,
`Microsecond64`,
`Nanosecond64`,
`DurationSecond`,
`DurationMillisecond`,
`DurationMicrosecond`,
`DurationNanosecond`,
`Decimal128`,
`Decimal256`,
`TimestampSecond`,
`TimestampMillisecond`,
`TimestampMicrosecond`,
`TimestampNanosecond`

Should I push whatever's done?

**Comment by ByteBaker** [comment]
@alamb I've merged the work for now. Please take a look.

I believe this covers most, if not all the non-nested types. I've also added the necessary docs, clearly mentioning the limitations of this macro.

**Comment by timsaucer** [comment]
Oh nice. FWIW, I don’t think we need to cover nested types. This is really just a convenience macro and primarily used for unit tests and examples, so it’s perfectly reasonable in my opinion that the scope is limited. For anything more the user can do what they’ve been doing until now.

I will try to review tomorrow or Monday unless someone gets to it first.

File: arrow-array/src/record_batch.rs [hunk 2]
```diff
@@ -411,6 +547,19 @@ impl RecordBatch {
     ///   ("b", b),
     /// ]);
     /// ```
+    /// Another way to quickly create a [`RecordBatch`] is to use the [`record_batch!`] macro,
```
Comment (alamb) [inline arrow-array/src/record_batch.rs:550]: 💯

**Review by alamb — APPROVED** [review]
Love it

**Comment by alamb** [comment]
Thanks @ByteBaker and @timsaucer

**Pull request merged**